<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Single Cell RNAseq • SPscrnaseq</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Single Cell RNAseq">
<meta property="og:description" content="SPscrnaseq">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-CCMXLYXWYV"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-CCMXLYXWYV');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPscrnaseq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/SPscrnaseq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/systemPipeR/SPscrnaseq/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Single Cell RNAseq</h1>
                        <h4 data-toc-skip class="author">Author:
FirstName LastName</h4>
            
            <h4 data-toc-skip class="date">Last update: 02 October,
2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/systemPipeR/SPscrnaseq/blob/HEAD/vignettes/SPscrnaseq.Rmd" class="external-link"><code>vignettes/SPscrnaseq.Rmd</code></a></small>
      <div class="hidden name"><code>SPscrnaseq.Rmd</code></div>

    </div>

    
    
<!--
Config css and r style
-->
<style type="text/css">
pre code {
white-space: pre !important;
overflow-x: scroll !important;
word-break: keep-all !important;
word-wrap: initial !important;
}
</style>
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div class="section level2">
<h2 id="about-the-template">About the template<a class="anchor" aria-label="anchor" href="#about-the-template"></a>
</h2>
<p>This section provides general description and how to use this single
cell RNAseq (scRNAseq) workflow. In the actual analysis report, this
section is usually <strong>removed</strong>.</p>
<p>This scRNAseq workflow template is based on the counted 10x scRNA
data. It means this workflow expect users have output from cell counting
programs, like cell ranger. If you have the raw sequencing data and
would like to count the cell gene counts, please use another workflow in
systemPipeR, such as <code>SPcellranger</code>.</p>
<p>This workflow does:</p>
<ol style="list-style-type: decimal">
<li>Read in single cell count data.</li>
<li>Basic stats on input data.</li>
<li>Create some basic QC on cell counting.</li>
<li>Normalization.</li>
<li>Find high variable genes.</li>
<li>Scaling.</li>
<li>dim reduction, PCA.</li>
<li>Clustering with tSNE, uMAP.</li>
<li>Find clustering markers (marker gene).</li>
<li>Find cell types.</li>
<li>Visualize cell types and clustering together.</li>
</ol>
<p>This template is modified from <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html" class="external-link">Seurat
Tutorial</a>.</p>
<p><img src="results/plotwf_scrna.jpg"></p>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Users want to provide here background information about the design of
their scRNAseq project.</p>
<p>This report describes the analysis of a scRNAseq project studying
drug …</p>
<div class="section level3">
<h3 id="experimental-design">Experimental design<a class="anchor" aria-label="anchor" href="#experimental-design"></a>
</h3>
<p>Typically, users want to specify here all information relevant for
the analysis of their scRNAseq study. This includes detailed
descriptions of files, experimental design, reference genome, gene
annotations, etc.</p>
</div>
</div>
<div class="section level2">
<h2 id="workflow-environment">Workflow environment<a class="anchor" aria-label="anchor" href="#workflow-environment"></a>
</h2>
<p><em><code>systemPipeR</code></em> workflows can be designed and built
from start to finish with a single command, importing from an R Markdown
file or stepwise in interactive mode from the R console.</p>
<p>This tutorial will demonstrate how to build the workflow in an
interactive mode, appending each step. The workflow is constructed by
connecting each step via <code>appendStep</code> method. Each
<code>SYSargsList</code> instance contains instructions needed for
processing a set of input files with a specific command-line or R
software and the paths to the corresponding outfiles generated by a
particular tool/step.</p>
<p>To create a Workflow within <em><code>systemPipeR</code></em>, we can
start by defining an empty container and checking the directory
structure:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://systempipe.org/" class="external-link">systemPipeR</a></span><span class="op">)</span></span>
<span><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html" class="external-link">SPRproject</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">sal</span></span></code></pre></div>
<div class="section level3">
<h3 id="load-packages">Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h3>
<p>This is an empty template that contains only one demo step. Refer to
our <a href="https://systempipe.org/sp/spr/spr_run/" class="external-link">website</a> for how
to add more steps. If you prefer a more enriched template, <a href="https://systempipe.org/sp/spr/templates/" class="external-link">read this page</a> for
other pre-configured templates.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu">crayon</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/crayon/man/crayon.html" class="external-link">blue</a></span><span class="op">$</span><span class="fu">bold</span><span class="op">(</span><span class="st">"To use this workflow, following R packages are expected:\n"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"'Seurat"</span>, <span class="st">"ggplot2"</span>, <span class="st">"ggpubr"</span>, <span class="st">"patchwork"</span>, <span class="st">"dplyr"</span>, <span class="st">"tibble"</span>,</span>
<span>    <span class="st">"readr'\n"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"', '"</span><span class="op">)</span></span>
<span><span class="co">### pre-end</span></span>
<span><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://systempipe.org/" class="external-link">systemPipeR</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/" class="external-link">ggpubr</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"load_packages"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="load-data">Load data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h3>
<p>In this example, the single cell data is preprocessed/filtered 10x
data from a healthy donor. Samples taken from peripheral blood
mononuclear cells (PBMCs), about 3000 cells.</p>
<p>Dataset can be downloaded with this link: <a href="https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz" class="external-link uri">https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz</a></p>
<p>If the link is not working, visit <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/1.1.0/pbmc3k?" class="external-link">10x
website</a> for updated links.</p>
<p>For your real data, please preprocess and put the dataset inside
<code>data</code> directory</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span></span>
<span>    <span class="co"># unzip the data</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/untar.html" class="external-link">untar</a></span><span class="op">(</span><span class="st">"data/pbmc3k_filtered_gene_bc_matrices.tar.gz"</span>, exdir <span class="op">=</span> <span class="st">"data"</span><span class="op">)</span></span>
<span>    <span class="co"># load data</span></span>
<span>    <span class="va">pbmc.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html" class="external-link">Read10X</a></span><span class="op">(</span>data.dir <span class="op">=</span> <span class="st">"data/filtered_gene_bc_matrices/hg19/"</span><span class="op">)</span></span>
<span>    <span class="co"># Use dim to see the size of dataset, example data has</span></span>
<span>    <span class="co"># 2700 cells x 32738 genes</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">pbmc.data</span><span class="op">)</span></span>
<span><span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"load_data"</span>, dependency <span class="op">=</span> <span class="st">"load_packages"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="simple-count-visulization">Simple count visulization<a class="anchor" aria-label="anchor" href="#simple-count-visulization"></a>
</h3>
<p>We can plot to see how many cells have good expressions.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span></span>
<span>    <span class="va">at_least_one</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">pbmc.data</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">count_p1</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="va">at_least_one</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">value</span><span class="op">)</span>, binwidth <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">pbmc.data</span><span class="op">)</span><span class="op">/</span><span class="fl">400</span><span class="op">)</span>,</span>
<span>        fill <span class="op">=</span> <span class="st">"#6b97c2"</span>, color <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/theme_pubr.html" class="external-link">theme_pubr</a></span><span class="op">(</span><span class="fl">16</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,</span>
<span>        <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Distribution of detected genes"</span>,</span>
<span>        x <span class="op">=</span> <span class="st">"Genes with at least one tag"</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">count_p2</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="fu">BiocGenerics</span><span class="fu">::</span><span class="fu">colSums</span><span class="op">(</span><span class="va">pbmc.data</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">value</span><span class="op">)</span>, bins <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">pbmc.data</span><span class="op">)</span><span class="op">/</span><span class="fl">50</span><span class="op">)</span>,</span>
<span>        fill <span class="op">=</span> <span class="st">"#6b97c2"</span>, color <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/theme_pubr.html" class="external-link">theme_pubr</a></span><span class="op">(</span><span class="fl">16</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,</span>
<span>        <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Expression sum per cell"</span>, x <span class="op">=</span> <span class="st">"Sum expression"</span><span class="op">)</span></span>
<span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="st">"results/count_plots.png"</span>, <span class="fl">1000</span>, <span class="fl">700</span><span class="op">)</span></span>
<span>    <span class="va">count_p1</span> <span class="op">+</span> <span class="va">count_p2</span> <span class="op">+</span> <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>tag_levels <span class="op">=</span> <span class="st">"A"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"count_plot"</span>, dependency <span class="op">=</span> <span class="st">"load_data"</span><span class="op">)</span></span></code></pre></div>
<p><img src="results/count_plots.png"></p>
</div>
<div class="section level3">
<h3 id="create-seurat-object">Create Seurat object<a class="anchor" aria-label="anchor" href="#create-seurat-object"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span></span>
<span>    <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">pbmc.data</span>, project <span class="op">=</span> <span class="st">"pbmc3k"</span>,</span>
<span>        min.cells <span class="op">=</span> <span class="fl">3</span>, min.features <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span>
<span>    <span class="co"># calculate mitochondria gene ratio</span></span>
<span>    <span class="va">sce</span><span class="op">[[</span><span class="st">"percent.mt"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html" class="external-link">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">sce</span>, pattern <span class="op">=</span> <span class="st">"^MT-"</span><span class="op">)</span></span>
<span><span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"create_seurat"</span>, dependency <span class="op">=</span> <span class="st">"load_data"</span><span class="op">)</span></span></code></pre></div>
<p>Filters applied here are:</p>
<ul>
<li>
<code>min.cells = 3</code>: for a single gene, it must be presented
in at least 3 cells.</li>
<li>
<code>min.features = 200</code>: for any cell, it must at least have
200 genes with readable gene counts.</li>
</ul>
<p>In your real data, you may want to adjust the filters for different
projects.</p>
</div>
<div class="section level3">
<h3 id="qc-of-cells">QC of cells<a class="anchor" aria-label="anchor" href="#qc-of-cells"></a>
</h3>
<p>Some indicators are important to show the quality of the cell batch,
for example</p>
<ul>
<li>N genes (features) per cell</li>
<li>N counts per cell</li>
<li>Mitochondria (mt) gene ratio.</li>
<li>…</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span></span>
<span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="st">"results/qc1.png"</span>, <span class="fl">700</span>, <span class="fl">700</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">sce</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"percent.mt"</span><span class="op">)</span>,</span>
<span>        ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">qc_p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeatureScatter.html" class="external-link">FeatureScatter</a></span><span class="op">(</span><span class="va">sce</span>, feature1 <span class="op">=</span> <span class="st">"nCount_RNA"</span>, feature2 <span class="op">=</span> <span class="st">"percent.mt"</span><span class="op">)</span></span>
<span>    <span class="va">qc_p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeatureScatter.html" class="external-link">FeatureScatter</a></span><span class="op">(</span><span class="va">sce</span>, feature1 <span class="op">=</span> <span class="st">"nCount_RNA"</span>, feature2 <span class="op">=</span> <span class="st">"nFeature_RNA"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="st">"results/qc2.png"</span>, <span class="fl">700</span>, <span class="fl">450</span><span class="op">)</span></span>
<span>    <span class="va">qc_p1</span> <span class="op">+</span> <span class="va">qc_p2</span> <span class="op">+</span> <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>tag_levels <span class="op">=</span> <span class="st">"A"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"qc_seurat"</span>, dependency <span class="op">=</span> <span class="st">"create_seurat"</span><span class="op">)</span></span></code></pre></div>
<p><img src="results/qc1.png"></p>
<p>Extreme numbers, outlines are likely low quality sequencing/bad
cells.</p>
<p><img src="results/qc2.png"></p>
<p>As shown on the figure above, A is the ratio of percent mt
<em>vs.</em> N count. Normally, no matter how many counts you can find
from a cell, the percent mt gene ratio should always be stable. For
cells with extremely high mt ratios, they should be removed. Looking at
x-axis, for a single cell, the N counts should also be inside a
reasonable range. Too low of counts indicates the cell gel bead is empty
during barcoding. Too high of counts indicates a single gel bead had
more than one cell during barcoding. Either these cases are bad, and
cells of these should be removed.</p>
<p>B is N genes per cell <em>vs.</em> N counts per cell. Normally, if as
the N counts increases, the N genes also increases. All cells should be
connected to approximately one line. If you see more than one line,
there may be problem with the sequencing, or cells are not from the same
batch.</p>
</div>
<div class="section level3">
<h3 id="filter-cells">Filter cells<a class="anchor" aria-label="anchor" href="#filter-cells"></a>
</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span></span>
<span>    <span class="co"># Based on the QC plots, please change the settings</span></span>
<span>    <span class="co"># accordingly</span></span>
<span>    <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">sce</span>, subset <span class="op">=</span> <span class="va">nFeature_RNA</span> <span class="op">&gt;</span> <span class="fl">200</span> <span class="op">&amp;</span> <span class="va">nFeature_RNA</span> <span class="op">&lt;</span></span>
<span>        <span class="fl">2500</span> <span class="op">&amp;</span> <span class="va">nCount_RNA</span> <span class="op">&lt;</span> <span class="fl">10000</span> <span class="op">&amp;</span> <span class="va">percent.mt</span> <span class="op">&lt;=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"filter_cells"</span>, dependency <span class="op">=</span> <span class="st">"create_seurat"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="normalization">Normalization<a class="anchor" aria-label="anchor" href="#normalization"></a>
</h3>
<p>For RNAseq data, usually log transformation and TPM is performed. For
scRNA, the normalization is</p>
<p><span class="math display">\[
Normalized\ count\ for\ a\ gene\ in\ a\ cell\ =log1p(\frac{Feature\
counts}{Total\ counts\ in\ a\ cell})\\
where\ log1p\ is\ ln(n+1)
\]</span></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span></span>
<span>    <span class="co"># scale.factor = 10000 is a convenient number for</span></span>
<span>    <span class="co"># plotting, so the normalized counts is ranged between</span></span>
<span>    <span class="co"># 0.xx to 10.</span></span>
<span>    <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">sce</span>, normalization.method <span class="op">=</span> <span class="st">"LogNormalize"</span>,</span>
<span>        scale.factor <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co"># compare counts before and after</span></span>
<span>    <span class="va">count_p_norm</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="fu">BiocGenerics</span><span class="fu">::</span><span class="fu">colSums</span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">data</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">value</span><span class="op">)</span>, bins <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">pbmc.data</span><span class="op">)</span><span class="op">/</span><span class="fl">50</span><span class="op">)</span>,</span>
<span>        fill <span class="op">=</span> <span class="st">"#6b97c2"</span>, color <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/theme_pubr.html" class="external-link">theme_pubr</a></span><span class="op">(</span><span class="fl">16</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,</span>
<span>        <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Total expression after normalization"</span>,</span>
<span>        x <span class="op">=</span> <span class="st">"Sum expression"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="st">"results/normalize_count_compare.png"</span>, <span class="fl">1000</span>, <span class="fl">700</span><span class="op">)</span></span>
<span>    <span class="va">count_p2</span> <span class="op">+</span> <span class="va">count_p_norm</span> <span class="op">+</span> <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>tag_levels <span class="op">=</span> <span class="st">"A"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"normalize"</span>, dependency <span class="op">=</span> <span class="st">"filter_cells"</span><span class="op">)</span></span></code></pre></div>
<p><img src="results/normalize_count_compare.png"></p>
<p>After the normalization, expression is close to a normal
distribution.</p>
</div>
<div class="section level3">
<h3 id="find-highly-variable-genes">Find highly variable genes<a class="anchor" aria-label="anchor" href="#find-highly-variable-genes"></a>
</h3>
<p>Some genes are highly variable among cells. These genes are the key
for the research. Here, some top-ranked variable genes are calculated
and will be used for downstream analysis.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span></span>
<span>    <span class="co"># 2000 is default</span></span>
<span>    <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">sce</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span>,</span>
<span>        nfeatures <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span>    <span class="co"># top 10 variable genes</span></span>
<span>    <span class="va">top10_var</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="fl">10</span><span class="op">)</span></span>
<span>    <span class="co"># plot the top 2000 variable genes and mark top 10</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="st">"results/variable_genes.png"</span>, <span class="fl">700</span>, <span class="fl">600</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/VariableFeaturePlot.html" class="external-link">VariableFeaturePlot</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>        <span class="fu"><a href="https://satijalab.org/seurat/reference/LabelPoints.html" class="external-link">LabelPoints</a></span><span class="op">(</span>points <span class="op">=</span> <span class="va">top10_var</span>, repel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/theme_pubr.html" class="external-link">theme_pubr</a></span><span class="op">(</span><span class="fl">16</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"find_var_genes"</span>, dependency <span class="op">=</span> <span class="st">"normalize"</span><span class="op">)</span></span></code></pre></div>
<p><img src="results/variable_genes.png"></p>
</div>
<div class="section level3">
<h3 id="scaling">Scaling<a class="anchor" aria-label="anchor" href="#scaling"></a>
</h3>
<p>To make the visualization within a reasonable range, scaling is
performed. This changes the expression data points to be centered at 0
with standard deviation of 1.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span></span>
<span>    <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">sce</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"scaling"</span>, dependency <span class="op">=</span> <span class="st">"find_var_genes"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="pca">PCA<a class="anchor" aria-label="anchor" href="#pca"></a>
</h3>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span></span>
<span>    <span class="co"># only use the top 2000 genes, first 50 PCs (default)</span></span>
<span>    <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">sce</span>, features <span class="op">=</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sce</span><span class="op">)</span>,</span>
<span>        npcs <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span>    <span class="co"># we can use following command to see first 5 genes in</span></span>
<span>    <span class="co"># each PC</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">pca</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, nfeatures <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"pca"</span>, dependency <span class="op">=</span> <span class="st">"scaling"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="plot-pca-results">Plot PCA results<a class="anchor" aria-label="anchor" href="#plot-pca-results"></a>
</h3>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span></span>
<span>    <span class="co"># plot PCA overview</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="st">"results/pca_overview.png"</span>, <span class="fl">500</span>, <span class="fl">500</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">sce</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="co"># plot top contributed genes in PC 1 and 2</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="st">"results/pca_loadings.png"</span>, <span class="fl">700</span>, <span class="fl">550</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/VizDimLoadings.html" class="external-link">VizDimLoadings</a></span><span class="op">(</span><span class="va">sce</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="co"># we can also use heatmap to show top genes in</span></span>
<span>    <span class="co"># different PCs</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="st">"results/pca_heatmap.png"</span>, <span class="fl">700</span>, <span class="fl">700</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/DimHeatmap.html" class="external-link">DimHeatmap</a></span><span class="op">(</span><span class="va">sce</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">6</span>, cells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">/</span><span class="fl">5</span>, balanced <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>        slot <span class="op">=</span> <span class="st">"scale.data"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"pca_plots"</span>, dependency <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span></code></pre></div>
<p><img src="results/pca_overview.png"></p>
<p><img src="results/pca_loadings.png"></p>
<p><img src="results/pca_heatmap.png"></p>
</div>
<div class="section level3">
<h3 id="find-optimal-pcs">Find optimal PCs<a class="anchor" aria-label="anchor" href="#find-optimal-pcs"></a>
</h3>
<p>After PCA, the dimension has reduced from tens of thousands
(features) to 50. Before, we do the clustering, we can further reduced
the PCs to only a few important ones, so the computation work can be
easier. How to choose the best number of PCs is hard to decide. Here,
the JackStraw plot and Elbow plot can be good ways to help us make the
decision.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span></span>
<span>    <span class="co"># for demo purposes, only a few replicates are used to</span></span>
<span>    <span class="co"># speed up the calculation in your real data, use</span></span>
<span>    <span class="co"># larger number like 100, etc.</span></span>
<span>    <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/JackStraw.html" class="external-link">JackStraw</a></span><span class="op">(</span><span class="va">sce</span>, num.replicate <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span>    <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScoreJackStraw.html" class="external-link">ScoreJackStraw</a></span><span class="op">(</span><span class="va">sce</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="st">"results/jackstraw.png"</span>, <span class="fl">660</span>, <span class="fl">750</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/JackStrawPlot.html" class="external-link">JackStrawPlot</a></span><span class="op">(</span><span class="va">sce</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="st">"results/elbow.png"</span>, <span class="fl">500</span>, <span class="fl">500</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/ElbowPlot.html" class="external-link">ElbowPlot</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"choose_pcs"</span>, dependency <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span></code></pre></div>
<p><img src="results/jackstraw.png"></p>
<p>There is a gap of values between PCs 1-6 and other PCs. Therefore PCs
1-6 are the most important. One could also include PCs 7-13 since they
have <span class="math inline">\(p\leq0.05\)</span>. One would need to
adjust the PC choice based on different datasets.</p>
<p><img src="results/elbow.png"></p>
<p>The Elbow plot also shows there is turning point around PC 7-8. All
other following PCs did not change too much.</p>
<p>It seems we should choose PC 6 or 7 as the cutoff point based on
plots above. However, Seurat documents recommend to combine methods
above with GSEA. Some cell populations may relate to genes in later PCs,
e.g. dendritic cell and NK aficionados are linked to PCs 12, 13.
Therefore, if your computational resources allow, it would be good to
include a few more PCs.</p>
<p>The Seurat documents recommend to repeat with 10, 15 or PCs. However,
in our experiences, the results did not change significantly.</p>
<p>In the downstream analysis, this template choose <strong>10</strong>
PCs.</p>
</div>
<div class="section level3">
<h3 id="clustering">Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h3>
<p>Seurat uses graph-based methods like KNN to find the clusters.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span></span>
<span>    <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">sce</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span>    <span class="co"># resolution 0.4-1.2 good for 3000 cells, if you have</span></span>
<span>    <span class="co"># more cells, increase the number will give you more</span></span>
<span>    <span class="co"># clusters</span></span>
<span>    <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">sce</span>, resolution <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"find_clusters"</span>, dependency <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualize-clusters-with-umaptsne">Visualize clusters with UMAP/tSNE<a class="anchor" aria-label="anchor" href="#visualize-clusters-with-umaptsne"></a>
</h3>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span></span>
<span>    <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">sce</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="st">"results/"</span><span class="op">)</span></span>
<span>    <span class="va">p_umap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">sce</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>    <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunTSNE.html" class="external-link">RunTSNE</a></span><span class="op">(</span><span class="va">sce</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span></span>
<span>    <span class="va">p_tsne</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">sce</span>, reduction <span class="op">=</span> <span class="st">"tsne"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="st">"results/plot_clusters.png"</span>, <span class="fl">1000</span>, <span class="fl">570</span><span class="op">)</span></span>
<span>    <span class="va">p_umap</span> <span class="op">+</span> <span class="va">p_tsne</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"plot_cluster"</span>, dependency <span class="op">=</span> <span class="st">"find_clusters"</span><span class="op">)</span></span></code></pre></div>
<p><img src="results/plot_clusters.png"></p>
</div>
<div class="section level3">
<h3 id="find-cluster-biomarkers">Find cluster biomarkers<a class="anchor" aria-label="anchor" href="#find-cluster-biomarkers"></a>
</h3>
<p>Find genes that represent different clusters.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span></span>
<span>    <span class="co"># find markers for every cluster compared to all</span></span>
<span>    <span class="co"># remaining cells, report only the positive ones</span></span>
<span>    <span class="va">sce.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers</a></span><span class="op">(</span><span class="va">sce</span>, only.pos <span class="op">=</span> <span class="cn">TRUE</span>, min.pct <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>        logfc.threshold <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></span>
<span>    <span class="va">sce.markers</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html" class="external-link">top_n</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">2</span>, wt <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span></span>
<span><span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"find_markers"</span>, dependency <span class="op">=</span> <span class="st">"find_clusters"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualize-markers">Visualize markers<a class="anchor" aria-label="anchor" href="#visualize-markers"></a>
</h3>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="st">"results/vlnplot.png"</span>, <span class="fl">600</span>, <span class="fl">600</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">sce</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MS4A1"</span>, <span class="st">"CD79A"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="st">"results/marker_features.png"</span>, <span class="fl">700</span>, <span class="fl">500</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">sce</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MS4A1"</span>, <span class="st">"GNLY"</span>, <span class="st">"CD3E"</span>, <span class="st">"CD14"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co"># plot top 10 DEG genes in each cluster as a heatmap</span></span>
<span>    <span class="va">top10_markers</span> <span class="op">&lt;-</span> <span class="va">sce.markers</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html" class="external-link">top_n</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span>, wt <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="st">"results/marker_heatmap.png"</span>, <span class="fl">1100</span>, <span class="fl">700</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/DoHeatmap.html" class="external-link">DoHeatmap</a></span><span class="op">(</span><span class="va">sce</span>, features <span class="op">=</span> <span class="va">top10_markers</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"plot_markers"</span>, dependency <span class="op">=</span> <span class="st">"find_markers"</span><span class="op">)</span></span></code></pre></div>
<p><img src="results/vlnplot.png"></p>
<p><img src="results/marker_features.png"></p>
<p><img src="results/marker_heatmap.png"></p>
</div>
<div class="section level3">
<h3 id="classify-cell-types">Classify cell types<a class="anchor" aria-label="anchor" href="#classify-cell-types"></a>
</h3>
<p>There are a few ways one can classify different clusters into
different cell types. The best way is you know what are the markers for
targeting cell types. CellMarker(<a href="http://biocc.hrbmu.edu.cn/CellMarker/" class="external-link uri">http://biocc.hrbmu.edu.cn/CellMarker/</a>) is a great source
to find markers of different cell types. If you do not know the markers,
use <code>singleR</code> package may be helpful. This package
automatically classify clusters into cell types. However, the accuracy
is not promised.</p>
<p>For example, if we know the markers of different cell types as
following:</p>
<table class="table">
<thead><tr class="header">
<th>Cluster</th>
<th>Markers</th>
<th>Cell Type</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>0</td>
<td>IL7R, CCR7</td>
<td>Naive CD4+ T</td>
</tr>
<tr class="even">
<td>1</td>
<td>CD14, LYZ</td>
<td>CD14+ Mono</td>
</tr>
<tr class="odd">
<td>2</td>
<td>IL7R, S100A4</td>
<td>Memory CD4+</td>
</tr>
<tr class="even">
<td>3</td>
<td>MS4A1</td>
<td>B</td>
</tr>
<tr class="odd">
<td>4</td>
<td>CD8A</td>
<td>CD8+ T</td>
</tr>
<tr class="even">
<td>5</td>
<td>FCGR3A, MS4A7</td>
<td>FCGR3A+ Mono</td>
</tr>
<tr class="odd">
<td>6</td>
<td>GNLY, NKG7</td>
<td>NK</td>
</tr>
<tr class="even">
<td>7</td>
<td>FCER1A, CST3</td>
<td>DC</td>
</tr>
<tr class="odd">
<td>8</td>
<td>PPBP</td>
<td>Platelet</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span></span>
<span>    <span class="va">new.cluster.ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Naive CD4 T"</span>, <span class="st">"CD14+ Mono"</span>, <span class="st">"Memory CD4 T"</span>,</span>
<span>        <span class="st">"B"</span>, <span class="st">"CD8 T"</span>, <span class="st">"FCGR3A+ Mono"</span>, <span class="st">"NK"</span>, <span class="st">"DC"</span>, <span class="st">"Platelet"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">new.cluster.ids</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span>    <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">RenameIdents</a></span><span class="op">(</span><span class="va">sce</span>, <span class="va">new.cluster.ids</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="st">"results/marker_labels.png"</span>, <span class="fl">700</span>, <span class="fl">700</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">sce</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"label_cell_type"</span>, dependency <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"plot_cluster"</span>,</span>
<span>    <span class="st">"find_markers"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="results/marker_labels.png"></p>
</div>
<div class="section level3">
<h3 id="workflow-session">Workflow session<a class="anchor" aria-label="anchor" href="#workflow-session"></a>
</h3>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"wf_session"</span>, dependency <span class="op">=</span> <span class="st">"label_cell_type"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="manage-the-workflow">Manage the workflow<a class="anchor" aria-label="anchor" href="#manage-the-workflow"></a>
</h2>
<p>To run the workflow, use <code>runWF</code> function. It executes all
the steps store in the workflow container. The execution will be on a
single machine without submitting to a queuing system of a computer
cluster.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runWF.html" class="external-link">runWF</a></span><span class="op">(</span><span class="va">sal</span>, run_step <span class="op">=</span> <span class="st">"mandatory"</span><span class="op">)</span>  <span class="co"># remove `run_step` to run all steps to include optional steps</span></span></code></pre></div>
<ul>
<li>To use complex workflow control options, such as parallelization,
subsetting samples, selecting steps, read the <a href="https://systempipe.org/sp/spr/sp_run/step_run/" class="external-link">documents</a> on
our website.</li>
<li>Explore <a href="https://systempipe.org/sp/spr/sp_run/sal_explore/" class="external-link">other details
of the workflow object</a>.</li>
<li>Create <a href="https://systempipe.org/sp/spr/sp_run/step_reports/" class="external-link">logs and
reports</a>.</li>
<li>
<a href="https://systempipe.org/sp/spr/sp_run/step_vis/" class="external-link">Visualize
the workflow</a>.</li>
</ul>
</div>
<div class="section level2">
<h2 id="about-the-workflow">About the workflow<a class="anchor" aria-label="anchor" href="#about-the-workflow"></a>
</h2>
<div class="section level3">
<h3 id="tools-used">Tools used<a class="anchor" aria-label="anchor" href="#tools-used"></a>
</h3>
<p>To check command-line tools used in this workflow, use
<code>listCmdTools</code>, and use <code>listCmdModules</code> to check
if you have a modular system.</p>
<p>The following code will print out tools required in your custom SPR
project in the report. In case you are running the workflow for the
first and do not have a project yet, or you just want to browser this
workflow, following code displays the tools required by default.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">".SPRproject"</span>, <span class="st">"SYSargsList.yml"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/eval.html" class="external-link">local</a></span><span class="op">(</span><span class="op">{</span></span>
<span>        <span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu">systemPipeR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html" class="external-link">SPRproject</a></span><span class="op">(</span>resume <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>        <span class="fu">systemPipeR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/listCmdTools.html" class="external-link">listCmdTools</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></span>
<span>        <span class="fu">systemPipeR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/listCmdTools.html" class="external-link">listCmdModules</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu">crayon</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/crayon/man/crayon.html" class="external-link">blue</a></span><span class="op">$</span><span class="fu">bold</span><span class="op">(</span><span class="st">"Tools and modules required by this workflow are:\n"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NA"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #0000BB; font-weight: bold;">Tools and modules required by this workflow are:</span></span></span>
<span><span class="co"><span style="color: #0000BB; font-weight: bold;">## </span>NA</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h3>
<p>This is the session information for rendering this report. To access
the session information of workflow running, check HTML report of
<code>renderLogs</code>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.6 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/Los_Angeles</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [6] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] BiocStyle_2.28.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] crayon_1.5.2          vctrs_0.6.2          </span></span>
<span><span class="co">##  [3] cli_3.6.1             knitr_1.42           </span></span>
<span><span class="co">##  [5] rlang_1.1.1           xfun_0.39            </span></span>
<span><span class="co">##  [7] stringi_1.7.12        purrr_1.0.1          </span></span>
<span><span class="co">##  [9] textshaping_0.3.6     jsonlite_1.8.4       </span></span>
<span><span class="co">## [11] glue_1.6.2            rprojroot_2.0.3      </span></span>
<span><span class="co">## [13] htmltools_0.5.5       formatR_1.14         </span></span>
<span><span class="co">## [15] ragg_1.2.5            sass_0.4.5           </span></span>
<span><span class="co">## [17] rmarkdown_2.21        evaluate_0.20        </span></span>
<span><span class="co">## [19] jquerylib_0.1.4       fastmap_1.1.1        </span></span>
<span><span class="co">## [21] lifecycle_1.0.3       yaml_2.3.7           </span></span>
<span><span class="co">## [23] memoise_2.0.1         bookdown_0.33        </span></span>
<span><span class="co">## [25] BiocManager_1.30.21.1 stringr_1.5.0        </span></span>
<span><span class="co">## [27] compiler_4.3.1        codetools_0.2-19     </span></span>
<span><span class="co">## [29] fs_1.6.2              rstudioapi_0.14      </span></span>
<span><span class="co">## [31] systemfonts_1.0.4     digest_0.6.31        </span></span>
<span><span class="co">## [33] R6_2.5.1              magrittr_2.0.3       </span></span>
<span><span class="co">## [35] bslib_0.4.2           tools_4.3.1          </span></span>
<span><span class="co">## [37] pkgdown_2.0.7         cachem_1.0.8         </span></span>
<span><span class="co">## [39] desc_1.4.2</span></span></code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Le Zhang, <a href="http://girke.bioinformatics.ucr.edu/" class="external-link">Thomas Girke</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
